### PR TITLE
Issue 2112: Adding duplicate count check to system test

### PR DESCRIPTION
**Pull request by shiveshr**
Signed-off-by: shivesh ranjan <shivesh.ranjan@gmail.com>

**Change log description**
* Changes AbstractFailoverTest.TestState to capture all events that are written and read back. 

* Records events written and read back and reports any anomalies.

* Reports txns left in committing or txns aborted at the time when test completes. 

**Purpose of the change**
Fixes #2113 

**What the code does**
Add further verification of duplicates to system test cases.

**How to verify it**
Run tests.

### CODE REVIEW

File: test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java [hunk 1]
```diff
@@ -97,22 +94,90 @@
...
+        }
+
+        void eventWritten(Long event) {
+            eventMap.putIfAbsent(event, 0);
+        }
+
+        void eventRead(Long event) {
+            eventMap.compute(event, (x, y) -> {
+                if (y == null) {
+                    return 1;
```
Comment (shrids) [inline test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: If y is null it implies that an event that was not written was read, right?
We should log this as an error.

File: test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java [hunk 2]
```diff
@@ -97,22 +94,90 @@
...
+        Long getNext() {
+            return eventData.getAndIncrement();
+        }
+
+        int getEventWrittenCount() {
+            return eventMap.size();
+        }
+
+        int getEventReadCount() {
+            return eventMap.values().stream().reduce(0, (x, y) -> x + y);
```
Comment (shrids) [inline test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: Nit: we can replace it with Integer::sum

File: test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java [hunk 3]
```diff
@@ -97,22 +94,90 @@
...
+        }
+
+        void printAnomalies() {
+            List<Long> notRead = eventMap.entrySet().stream().filter(x -> x.getValue() == 0)
+                    .map(Map.Entry::getKey).collect(Collectors.toList());
+            if (notRead.size() > 0) {
+                log.error("anomalies, unread events => {}", notRead);
+            }
+
+            Map<Long, Integer> duplicates = eventMap.entrySet().stream().filter(x -> x.getValue() == 0)
```
Comment (shrids) [inline test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: if (value is > 1)) then it is a duplicate.

File: test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java (hunk 1 above)
Comment (shiveshr) [reply test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: No before the write ack comes we could have read the event back..

File: test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java (hunk 3 above)
Comment (shiveshr) [reply test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: Oops

File: test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java [hunk 4]
```diff
@@ -97,22 +94,90 @@
...
+            return eventMap.values().stream().reduce(0, (x, y) -> x + y);
+        }
+
+        void printAnomalies() {
+            List<Long> notRead = eventMap.entrySet().stream().filter(x -> x.getValue() == 0)
+                    .map(Map.Entry::getKey).collect(Collectors.toList());
+            if (notRead.size() > 0) {
+                log.error("anomalies, unread events => {}", notRead);
+            }
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: does it do any good if we print null events also ?

File: test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java [hunk 5]
```diff
@@ -97,22 +94,90 @@
...
+            Map<Long, Integer> duplicates = eventMap.entrySet().stream().filter(x -> x.getValue() > 1)
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+            if (duplicates.size() > 0) {
+                log.error("anomalies, duplicate events with count => {}", duplicates);
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: Also, can you please capitalise the first letter of the log statement, here and everywhere

File: test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java (hunk 4 above)
Comment (shiveshr) [reply test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: what are null events?

Comment (Prabha-Veerubhotla) [reply test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: these ?
https://github.com/pravega/pravega/pull/2113/files#diff-7c588e31332a0f89d89d04375f736605R115

Comment (shiveshr) [reply test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: These are not going to result in null events.. 
you cant use null as a key for a map anyway :) 
what this is doing is map.compute(key, (x,y) -> {})
this computes and updates the map based on if an entry against key is present or not. 
y will be null if map did not have an entry for this key.

File: test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java [hunk 6]
```diff
@@ -97,22 +95,90 @@
...
+
+        int getEventWrittenCount() {
+            return eventMap.size();
+        }
+
+        int getEventReadCount() {
+            return eventMap.values().stream().mapToInt(Integer::intValue).sum();
+        }
+
+        void printAnomalies() {
```
Comment (fpj) [inline test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: The idea of having a method that centralizes the verification that nothing has gone wrong is great. I wonder if we should strengthen it, though. We could call it `checkForAnomalies` and fail the test in the case anything fails. Currently, we are just logging.

Comment (shiveshr) [reply test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java]: done

**Pull request merged**